samcli/commands/destroy/__init__.py

```diff
@@ -0,0 +1,215 @@
+        retain_resources = []
+    paginator = client.get_paginator('describe_stack_events')
+    response_iterator = paginator.paginate(
+        StackName=stack_name
+    )
+    for iterator in response_iterator:
+        events = [event for event in iterator.get("StackEvents") if
+                  event.get("ResourceStatus") == "DELETE_FAILED"]
+        for event in events:
+            logical_id = event.get("LogicalResourceId")
+            if logical_id not in retain_resources:
+                click.secho(
+                    "The logicalId {} of the resource in the stack {} must be included in retain_resource since the "
+                    "deletion failed".format(logical_id, stack_name), fg="red")
+                sys.exit(1)
+
+
+def do_cli(ctx, stack_name, retain_resources=None, role_arn=None, client_request_token=None, wait=None, wait_time=None,
+           force=False):
+    """
+    Implementation of the ``cli`` method, just separated out for unit testing purposes
+    """
+    if not force:
+        click.confirm('Are you sure you want to delete the stack {}?'.format(stack_name), default=True, abort=True)
+    cfn = boto3.client('cloudformation', region_name='us-west-1')
+
+    verify_stack_exists(cfn, stack_name)
+    verify_stack_retain_resources(cfn, stack_name, retain_resources)
+    if not force:
+        check_nested_stack(cfn, stack_name)
+
+    args = {'RoleARN': role_arn, 'ClientRequestToken': client_request_token, 'RetainResources': retain_resources}
+
+    # Filters the args dictionary so that no argument with type `None` is passed in. This is because deleting a stack
+    # with boto3 only accepts non `None` arguments.
+    args = {k: v for k, v in args.items() if v is not None}
+
+    try:
+        cfn.delete_stack(StackName=stack_name, **args)
+    except ClientError as e:
+        if "TerminationProtection" in e.response["Error"]["Message"]:
+            msg = 'The stack {stack_name} has TerminationProtection turned on. Disable it on the aws console at ' + \
+                  'https://us-west-1.console.aws.amazon.com/cloudformation/home \n or ' + \
+                  'run aws cloudformation update-termination-protection --stack-name {stack_name} ' + \
+                  '--no-enable-termination-protection'
+            click.secho(msg
+                        .format(stack_name=stack_name), fg="red")
+        elif "AccessDeniedException" in e.response["Error"]["Message"]:
+            click.secho(
+                'The user account does not have access to delete the stack. \n' +
+                'Please update the resources required to delete the stack and the required user policies.', fg="red")
+        else:
+            click.secho("Failed to destroy Stack: {}".format(str(e.response["Error"]["Message"])), fg="red")
+
+        sys.exit(1)
+
+    if wait or wait_time:
+        waiter = cfn.get_waiter('stack_delete_complete')
+        try:
+            delay = 15
+            waiter.wait(StackName=stack_name,
```

Comment:
is this a poller implementation that has been surfaced?

Reply:
Ye. It uses a poller behind the scenes